File: C#/MarketServer/Market.cs
using System;
using System.Collections.Generic;

namespace MarketServer
{
    public class Market
    {
        private int stockHolderID = -1;
        public List<int> connectedTraders = new List<int>();

        public bool GiveStockTo(int traderID){
            if (connectedTraders.Contains(traderID)){
                stockHolderID = traderID;
                Console.WriteLine("Share has been given to " + stockHolderID);
                return true;
            } else {
                Console.WriteLine("Cannot give stock to " + traderID);
                return false;
            }
        }

        public int GetStockHolderID() {
            return stockHolderID;
        }

        public void NoTradersLeft(){
            stockHolderID = -1;
            Console.WriteLine("No traders left - waiting to reassign the stock to the next customer that connects");
        }

        public void AddToConnectedTraders(int traderID){
            connectedTraders.Add(traderID);
        }

        public void RemoveFromConnectedTraders(int traderID){

            connectedTraders.Remove(traderID);
            Console.WriteLine("Trader " + traderID + " was removed from the connected traders list");
            if (stockHolderID == traderID){
                if (connectedTraders.Count > 1){
                    GiveStockTo(connectedTraders[connectedTraders.Count-1]);
                    Console.WriteLine("Stock was automatically given to " + connectedTraders[connectedTraders.Count-1]);
                } else {
                    NoTradersLeft();
                }
            }
        }
    }
}

File: Java/MarketServer/src/Market.java
import java.util.ArrayList;
import java.util.List;

public class Market {

    private int shareHolderID = -1;
    public List<Integer> connectedCustomers = new ArrayList<>();

    public boolean giveShareTo(int customerID){
        if (connectedCustomers.contains(customerID)){
            shareHolderID = customerID;
            System.out.println("Stock has been given to Trader " + shareHolderID);
            return true;
        } else {
            System.out.println("Cannot give stock to Trader " + customerID + " because they are not connected");
            return false;
        }
    }
    public int getStockHolderID() {
        return shareHolderID;
    }

    public void noTradersLeft(){
        shareHolderID = -1;
    }

    public void addToConnectedCustomers(int customerID){
        connectedCustomers.add(customerID);
    }

    public void removeFromConnectedCustomers(int customerID){
        System.out.println("The current assigned stockholder is " + shareHolderID);
        connectedCustomers.remove(Integer.valueOf(customerID));
        System.out.println("Removed Trader " + customerID + " from list of connected traders");
        if (shareHolderID == customerID){
            if (connectedCustomers.size() > 1) {
                giveShareTo(connectedCustomers.get(connectedCustomers.size()-1));
                System.out.println("Stockholder (Trader " + customerID + ") disconnected, so stock was automatically given to Trader " + connectedCustomers.get(connectedCustomers.size()-1));
            } else {
                noTradersLeft();
            }
        }

    }


}
